File: HW1_Particle_Filter/python/MotionModel.py
import sys
import numpy as np
import math


class MotionModel:

    """
    References: Thrun, Sebastian, Wolfram Burgard, and Dieter Fox. Probabilistic robotics. MIT press, 2005.
    [Chapter 5]
    """

    def __init__(self):

        """
        TODO : Initialize Motion Model parameters here
        """
        self.alpha1 = 0.05
        self.alpha2 = 0.04
        self.alpha3 = 0.03
        self.alpha4 = 0.04

    def update(self, u_t0, u_t1, x_t0):
        """
        param[in] u_t0 : particle state odometry reading [x, y, theta] at time (t-1) [odometry_frame]   
        param[in] u_t1 : particle state odometry reading [x, y, theta] at time t [odometry_frame]
        param[in] x_t0 : particle state belief [x, y, theta] at time (t-1) [world_frame]
        param[out] x_t1 : particle state belief [x, y, theta] at time t [world_frame]
        """

        """
        TODO : Add your code here
        """
        rot1 = math.atan2(u_t1[1] - u_t0[1], u_t1[0] - u_t0[0]) - u_t0[2]
        trans = math.hypot(u_t1[1] - u_t0[1], u_t1[0] - u_t0[0])
        rot2 = u_t1[2] - u_t0[2] - rot1
        
        stddev_rot1 = math.hypot(self.alpha1 * rot1, self.alpha2 * trans)
        stddev_trans = self.alpha4 * math.hypot(rot1, rot2)
        stddev_trans = math.hypot(stddev_trans, self.alpha3 * trans)
        stddev_rot2 = math.hypot(self.alpha1 * rot2, self.alpha2 * trans)
        
        rot1_noisy = np.random.normal(loc=rot1, scale=stddev_rot1)
        trans_noisy = np.random.normal(loc=trans, scale=stddev_trans)
        rot2_noisy = np.random.normal(loc=rot2, scale=stddev_rot2)
        
        x_t1 = x_t0
        x_t1[2] += rot1_noisy
        x_t1 += np.array([trans_noisy * math.cos(x_t1[2]), trans_noisy * math.sin(x_t1[2]), rot2_noisy])
        return x_t1


if __name__ == "__main__":
    pass

File: HW1_Particle_Filter/cpp/MotionModel.cpp
/**
    16833 Assignment 1: Particle Filter (re-written)
    @file MotionModel.cpp
    @author Lin, Zhaozhi
    @date Dec 2019
*/


#include "ParticleFilter.hpp"

MotionModel::MotionModel(const double& alpha1, 
    const double& alpha2, 
    const double& alpha3, 
    const double& alpha4) {
    this->alpha1 = alpha1;
    this->alpha2 = alpha2;
    this->alpha3 = alpha3;
    this->alpha4 = alpha4;
    this->rng = cv::RNG(time(nullptr));
}

cv::Vec3d MotionModel::computeTransformation(const cv::Vec3d& pose0, const cv::Vec3d& pose1) {
    const cv::Vec3d diff = pose1 - pose0;
    double rot1 = std::atan2(diff[1], diff[0]) - pose0[2];
    double trans = std::hypot(diff[0], diff[1]);
    double rot2 = diff[2] - rot1;
    return cv::Vec3d(rot1, trans, rot2);
}

cv::Vec3d MotionModel::computeStandardDeviation(const cv::Vec3d& transformations) {

    double rot1 = transformations[0];
    double trans = transformations[1];
    double rot2 = transformations[2];

    double stddevRot1 = std::hypot(this->alpha1 * rot1, this->alpha2 * trans);
    double stddevTrans = this->alpha4 * std::hypot(rot1, rot2);
    stddevTrans = std::hypot(stddevTrans, this->alpha3 * trans);
    double stddevRot2 = std::hypot(this->alpha1 * rot2, this->alpha2 * trans);
    return cv::Vec3d(stddevRot1, stddevTrans, stddevRot2);
}

void MotionModel::update(std::vector<cv::Vec4d>& particles, const cv::Vec3d& pose0, const cv::Vec3d& pose1) {
    auto transformations = this->computeTransformation(pose0, pose1);
    auto stddev = this->computeStandardDeviation(transformations);

    double rot1 = transformations[0];
    double trans = transformations[1];
    double rot2 = transformations[2];

    double stddevRot1 = stddev[0];
    double stddevTrans = stddev[1];
    double stddevRot2 = stddev[2];

    cv::parallel_for_(cv::Range(0, particles.size()), [&](const cv::Range& range) {
        for (auto index = range.start; index < range.end; ++index) {
            double rot1Noisy = rot1 + rng.gaussian(stddevRot1);
            double transNoisy = trans + rng.gaussian(stddevTrans);
            double rot2Noisy = rot2 + rng.gaussian(stddevRot2);

            particles[index][2] += rot1Noisy;
            particles[index][0] += transNoisy * std::cos(particles[index][2]);
            particles[index][1] += transNoisy * std::sin(particles[index][2]);
            particles[index][2] += rot2Noisy;
        }
    });

    /* Sequential version of above block
    for (auto& particle : particles) {

        double rot1Noisy = rot1 + rng.gaussian(stddevRot1);
        double transNoisy = trans + rng.gaussian(stddevTrans);
        double rot2Noisy = rot2 + rng.gaussian(stddevRot2);

        particle[2] += rot1Noisy;
        particle[0] += transNoisy * std::cos(particle[2]);
        particle[1] += transNoisy * std::sin(particle[2]);
        particle[2] += rot2Noisy;
    }
    */
}